File: codility/FrogRiverOne/Solution2.java
/**
 * Author: Liang Lin
 * Email: lianglin@outlook.com
 * Date: 2016-03-26
 */
import java.util.BitSet;
Class Solution {
    public int solution(int X, int[] A) {
        //solved using hash table
        int counter = 0;
        int n = A.length;
        BitSet B = new BitSet(X+1);
        for (int i=0; i<n; i++) {
            if(A[i] <= X && !B.get(A[i])) {
                B.set(A[i]);
                counter++;
                if(counter >= X) {
                    return i;
                }
            }
        }
        return -1;
    }
}

File: codility/3.TimeComplexity/PermMissingElem/solution2.py
# Author: lialin
# Email: lianglin@outlook.com
# Date: 2016-04-06
class Solution:
  def solution(self, A):
    n = len(A)
    B = [False] * n
    for val in A:
      if val <=n and val > 0:
	B[val-1] = True
    for i in xrange(n):
      if B[i] == False:
	return i+1
    return n+1
  
